[REVIEW] NEXUS-5468: HTTP connection leak in case of LSEx

cstamas opened the pull request:
Fix for LRS to always clean up PreparedContentLocator instances
and a UT verifying this happens in both case, in "norma" but
also in "early unexpected" exception case.

Issue:
https://issues.sonatype.org/browse/NEXUS-5468

CI
https://bamboo.zion.sonatype.com/browse/NXO-OSSF2-1

cstamas committed:
NEXUS-5468: HTTP connection leak in case of LSEx

Fix for LRS to always clean up PreparedContentLocator instances
and a UT verifying this happens in both case, in "norma" but
also in "early unexpected" exception case.

cstamas committed:
NEXUS-5468: Applying original parts

Instead of introducing new custom method. PreparedContentLocator
made plain Closeable. Using guava Closeables to
make it closed.

kellyrob99 commented:
+1

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/storage/local/fs/DefaultFSLocalRepositoryStorage.java:
```diff
@@ -328,42 +330,67 @@ public AbstractStorageItem retrieveItem( Repository repository, ResourceStoreReq
+            // get closed irrelevant of the actual outcome. If all went right, stream was already closed,
+            // and we will be "punished" by one extra (redundant) call to Closeable#close().
+            if ( originalContentLocator != null
+                && ( originalContentLocator instanceof PreparedContentLocator ) )
```
fyi, instanceof implies !=null

ifedorenko commented:
-1

The code is specific to PreparedContentLocator, however there is no guarantee this other ContentLocator locators do not also leak opened InputStream instances, either directly or through AbstractWrappingContentLocator. 

One way is to completely rework LocalRepositoryStorage/FSPeer API, which I believe we'll need to do in the long term, but this is way too big a change for 2.5.

In the short term, I suggest 1) change DefaultFSLocalRepositoryStorage#storeItem to close any Closable ContentLocator implementation and 2) make AbstractWrappingContentLocator implement Closeable and close wrapped location if it implements Closeable.

cstamas commented:
Agreed, will change this. I was also unease about being tied to `PreparedContentLocator`.

cstamas committed:
NEXUS-5468: Generalizing the solution

To not be specific for PreparedContentLocator, but instead
make them proper "Closeable"s. Same for FS RLS, it cares about
Closeable, not about given ContentLocator implementation.

cstamas commented:
Igor's comments applied.
